docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+This corresponds to a transformation used in Distributed Optimizers which flattens the data and shards it along the data-parallel domain.
+
+Extra flattening comes with an efficiency challenge during checkpoint resharding.
+Since flattening is applied after the global tensors is sharded into the grid of local chunks, loading after resharding requires accessing incontiguous data fragments.
+An example solution for that is implemented in the `dist_checkpointing/strategies/resharding.py` module and involves saving the flattened tensor with a different global shape than the original one.
+
+Example: For a global tensor [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11]] with sharding by TP (tensor-parallel) over the second axis, here are the local shards if TP=2:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [[0, 1, 2], [6, 7, 8]]
+   * - 1
+     - [[3, 4, 5], [9, 10, 11]]
+
+After flattening and sharding by e.g. DP=3:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [0, 1]
+   * - 2
+     - [2, 6]
+   * - 4
+     - [7, 8]
+   * - 1
+     - [3, 4]
+   * - 3
+     - [5, 9]
+   * - 5
+     - [10, 11]
+
+The same tensor after sharding by TP=6, flattening and sharding by DP=1:
+
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [0, 6]
+   * - 1
+     - [1, 7]
+   * - 2
+     - [2, 8]
+   * - 3
+     - [3, 9]
+   * - 4
+     - [4, 10]
+   * - 5
+     - [5, 11]
+
+
+Arbitrary transformations
+-------------------------
+The way to apply arbitrary transformations to the tensors during saving and loading is with ShardedTensorFactory, which allows to define such transformations as a function that can be reapplied to any ShardedTensor (in particular, a ShardedTensor representing optimizer states).
+Such "build" function is also tied to a "merge" function that can apply an inverse transformation during loading.
+
+If handling an optimizer state is not required, such transformation could be also applied directly during sharded state dict creation.
+In order to apply such transformation both to model and optimizer parameters in a consistent manner, it's necessary to encode them as factory functions (with original model parameter as the `data` input so that the optimizer params can be properly mapped to model ShardedTensors).
+
+Note that implementing some transformations might be challenging or impossible while supporting flattening for a Distributed Optimizer case.
+For example, if the model weights are supposed to be transposed in the checkpoint, it's almost impossible to implement a performant factory function that is capable of transposing a flattened and sliced tensor, because the flattening and slicing should happen in the transposed dimension.
```

Comment:
Keep in mind that implementing certain transformations can be quite challenging or even impossible while still supporting flattening, especially in the case of Distributed Optimizers. For instance, consider a scenario where the model weights need to be transposed in the checkpoint. Creating a performant factory function capable of transposing a flattened and sliced tensor becomes nearly impossible because the flattening and slicing operations should occur in the transposed dimension.